File: projects/nomikai/android/app/src/main/java/com/example/nomikai/data/db/NotificationRecord.kt
package com.example.nomikai.data.db

import androidx.room.*
import kotlinx.coroutines.flow.Flow

// ─────────────────────────────────────────────
//  Entity
// ─────────────────────────────────────────────
/**
 * 受信プッシュ通知のキャッシュActiveRecordモデル。
 * FCM受信時にローカルDBへ保存し、通知一覧画面はDBから読み込む。
 *
 * ```kotlin
 * // FCM受信時に保存
 * NotificationRecord(
 *     id        = uuid,
 *     userId    = currentUserId,
 *     inviteId  = data["inviteId"],
 *     title     = remoteMessage.notification?.title ?: "",
 *     body      = remoteMessage.notification?.body ?: "",
 *     dataJson  = gson.toJson(data)
 * ).save()
 *
 * // 一覧取得
 * val all    = NotificationRecord.allForUser(userId)
 * val unread = NotificationRecord.unreadForUser(userId)
 * val count  = NotificationRecord.unreadCount(userId)
 *
 * // 既読
 * notification.markRead()
 *
 * // 全件既読
 * NotificationRecord.markAllRead(userId)
 * ```
 */
@Entity(
    tableName = "notifications",
    indices = [
        Index(value = ["user_id", "created_at"]),
        Index(value = ["user_id", "is_read"])
    ]
)
data class NotificationRecord(
    @PrimaryKey val id: String,
    @ColumnInfo(name = "user_id") val userId: String,
    @ColumnInfo(name = "invite_id") val inviteId: String? = null,
    val title: String,
    val body: String,
    @ColumnInfo(name = "data_json") val dataJson: String? = null,
    @ColumnInfo(name = "is_read") val isRead: Int = 0,
    @ColumnInfo(name = "created_at") val createdAt: Long = System.currentTimeMillis()
) : ActiveRecord() {

    // ─── 算出プロパティ ────────────────────────────

    val isUnread: Boolean get() = isRead == 0

    // ─── インスタンスメソッド ───────────────────────

    override suspend fun save(): Long =
        DatabaseHolder.db.notificationDao().upsert(this)

    override suspend fun delete(): Int =
        DatabaseHolder.db.notificationDao().delete(this)

    /**
     * この通知を既読にする。
     * @return 既読状態になった新しいインスタンス
     */
    suspend fun markRead(): NotificationRecord {
        val updated = copy(isRead = 1)
        DatabaseHolder.db.notificationDao().markRead(id)
        return updated
    }

    // ─── クラスメソッド (companion) ────────────────

    companion object {
        private val dao get() = DatabaseHolder.db.notificationDao()

        /** ユーザーの全通知を新しい順で取得する */
        suspend fun allForUser(userId: String): List<NotificationRecord> =
            dao.findAllByUser(userId)

        /** ユーザーの未読通知のみ取得する */
        suspend fun unreadForUser(userId: String): List<NotificationRecord> =
            dao.findUnreadByUser(userId)

        /** 未読件数を取得する */
        suspend fun unreadCount(userId: String): Int =
            dao.countUnread(userId)

        /** ユーザーの通知をFlowで監視する */
        fun observeForUser(userId: String): Flow<List<NotificationRecord>> =
            dao.observeByUser(userId)

        /** 未読件数をFlowで監視する */
        fun observeUnreadCount(userId: String): Flow<Int> =
            dao.observeUnreadCount(userId)

        /** ユーザーの全通知を既読にする */
        suspend fun markAllRead(userId: String) =
            dao.markAllRead(userId)

        /** 30件を超えた古い通知を削除する（ストレージ節約） */
        suspend fun pruneOld(userId: String, keepCount: Int = 50) =
            dao.deleteOldExceeding(userId, keepCount)
    }
}

// ─────────────────────────────────────────────
//  DAO (非公開 - ActiveRecordの内部実装)
// ─────────────────────────────────────────────
@Dao
interface NotificationDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun upsert(record: NotificationRecord): Long

    @Delete
    suspend fun delete(record: NotificationRecord): Int

    @Query("SELECT * FROM notifications WHERE user_id = :userId ORDER BY created_at DESC")
    suspend fun findAllByUser(userId: String): List<NotificationRecord>

    @Query("SELECT * FROM notifications WHERE user_id = :userId AND is_read = 0 ORDER BY created_at DESC")
    suspend fun findUnreadByUser(userId: String): List<NotificationRecord>

    @Query("SELECT COUNT(*) FROM notifications WHERE user_id = :userId AND is_read = 0")
    suspend fun countUnread(userId: String): Int

    @Query("SELECT * FROM notifications WHERE user_id = :userId ORDER BY created_at DESC")
    fun observeByUser(userId: String): Flow<List<NotificationRecord>>

    @Query("SELECT COUNT(*) FROM notifications WHERE user_id = :userId AND is_read = 0")
    fun observeUnreadCount(userId: String): Flow<Int>

    @Query("UPDATE notifications SET is_read = 1 WHERE id = :id")
    suspend fun markRead(id: String)

    @Query("UPDATE notifications SET is_read = 1 WHERE user_id = :userId AND is_read = 0")
    suspend fun markAllRead(userId: String)

    /**
     * 新しい順で keepCount 件を残し、古いレコードを削除する。
     * サブクエリでrowid上位N件のIDを特定し、それ以外を削除。
     */
    @Query("""
        DELETE FROM notifications 
        WHERE user_id = :userId 
        AND id NOT IN (
            SELECT id FROM notifications 
            WHERE user_id = :userId 
            ORDER BY created_at DESC 
            LIMIT :keepCount
        )
    """)
    suspend fun deleteOldExceeding(userId: String, keepCount: Int)
}

File: projects/nomikai/ios/NomikaiApp/DB/Records/NotificationRecord.swift
import Foundation
import GRDB
import Combine

// ─────────────────────────────────────────────────────────────────────────
//  NotificationRecord - プッシュ通知のローカルキャッシュ
//
//  使い方:
//    // FCM受信時に保存
//    try NotificationRecord(
//        id: uuid, userId: uid, inviteId: data["inviteId"],
//        title: title, body: body, dataJson: jsonStr
//    ).save()
//
//    // 一覧取得
//    let all    = try NotificationRecord.allForUser(userId: uid)
//    let unread = try NotificationRecord.unreadForUser(userId: uid)
//    let count  = try NotificationRecord.unreadCount(userId: uid)
//
//    // 既読
//    try notification.markRead()
//    try NotificationRecord.markAllRead(userId: uid)
//
//    // Flow で監視
//    NotificationRecord.observeForUser(userId: uid)
//      .sink { notifications in ... }
// ─────────────────────────────────────────────────────────────────────────

struct NotificationRecord: ActiveRecord, Codable, Equatable {
    typealias ID = String

    var id: String
    var userId: String
    var inviteId: String?
    var title: String
    var body: String
    var dataJson: String?   // JSON文字列
    var isRead: Int         // 0=未読, 1=既読
    var createdAt: Int64

    enum Columns {
        static let id        = Column(CodingKeys.id)
        static let userId    = Column(CodingKeys.userId)
        static let inviteId  = Column(CodingKeys.inviteId)
        static let title     = Column(CodingKeys.title)
        static let body      = Column(CodingKeys.body)
        static let dataJson  = Column(CodingKeys.dataJson)
        static let isRead    = Column(CodingKeys.isRead)
        static let createdAt = Column(CodingKeys.createdAt)
    }

    enum CodingKeys: String, CodingKey {
        case id, title, body
        case userId    = "user_id"
        case inviteId  = "invite_id"
        case dataJson  = "data_json"
        case isRead    = "is_read"
        case createdAt = "created_at"
    }

    static var databaseTableName: String { "notifications" }
    static var persistenceConflictPolicy = PersistenceConflictPolicy(
        insert: .replace,
        update: .replace
    )

    init(
        id: String = UUID().uuidString,
        userId: String,
        inviteId: String? = nil,
        title: String,
        body: String,
        dataJson: String? = nil,
        isRead: Int = 0,
        createdAt: Int64 = .now
    ) {
        self.id        = id
        self.userId    = userId
        self.inviteId  = inviteId
        self.title     = title
        self.body      = body
        self.dataJson  = dataJson
        self.isRead    = isRead
        self.createdAt = createdAt
    }
}

// ─── 算出プロパティ ───────────────────────────────────────────────────────
extension NotificationRecord {
    var isUnread: Bool { isRead == 0 }
    var createdDate: Date { Date(timeIntervalSince1970: Double(createdAt) / 1000) }

    /// dataJson をデコードして通知データを返す。
    var notificationData: [String: String]? {
        guard let json = dataJson,
              let data = json.data(using: .utf8),
              let dict = try? JSONDecoder().decode([String: String].self, from: data)
        else { return nil }
        return dict
    }
}

// ─── インスタンスメソッド ─────────────────────────────────────────────────
extension NotificationRecord {
    /// この通知を既読にする。
    @discardableResult
    func markRead() throws -> NotificationRecord {
        var updated = self
        updated.isRead = 1
        return try updated.save()
    }
}

// ─── クラスメソッド（ファインダー） ─────────────────────────────────────
extension NotificationRecord {
    /// ユーザーの全通知を新しい順で取得する。
    static func allForUser(userId: String) throws -> [NotificationRecord] {
        try DatabaseHolder.shared.db.read { db in
            try NotificationRecord
                .filter(Columns.userId == userId)
                .order(Columns.createdAt.desc)
                .fetchAll(db)
        }
    }

    /// ユーザーの未読通知を取得する。
    static func unreadForUser(userId: String) throws -> [NotificationRecord] {
        try DatabaseHolder.shared.db.read { db in
            try NotificationRecord
                .filter(Columns.userId == userId && Columns.isRead == 0)
                .order(Columns.createdAt.desc)
                .fetchAll(db)
        }
    }

    /// 未読件数を取得する。
    static func unreadCount(userId: String) throws -> Int {
        try DatabaseHolder.shared.db.read { db in
            try NotificationRecord
                .filter(Columns.userId == userId && Columns.isRead == 0)
                .fetchCount(db)
        }
    }

    /// 全通知を既読にする。
    static func markAllRead(userId: String) throws {
        try DatabaseHolder.shared.db.write { db in
            try db.execute(
                sql: "UPDATE notifications SET is_read = 1 WHERE user_id = ? AND is_read = 0",
                arguments: [userId]
            )
        }
    }

    /// 通知一覧を Combine Publisher で監視する。
    static func observeForUser(userId: String) -> AnyPublisher<[NotificationRecord], Error> {
        ValueObservation
            .tracking { db in
                try NotificationRecord
                    .filter(Columns.userId == userId)
                    .order(Columns.createdAt.desc)
                    .fetchAll(db)
            }
            .publisher(in: DatabaseHolder.shared.db, scheduling: .immediate)
            .eraseToAnyPublisher()
    }

    /// 未読件数を Combine Publisher で監視する。
    static func observeUnreadCount(userId: String) -> AnyPublisher<Int, Error> {
        ValueObservation
            .tracking { db in
                try NotificationRecord
                    .filter(Columns.userId == userId && Columns.isRead == 0)
                    .fetchCount(db)
            }
            .publisher(in: DatabaseHolder.shared.db, scheduling: .immediate)
            .eraseToAnyPublisher()
    }

    /// 古い通知を削除して keepCount 件だけ残す。
    static func pruneOld(userId: String, keepCount: Int = 50) throws {
        try DatabaseHolder.shared.db.write { db in
            try db.execute(sql: """
                DELETE FROM notifications
                WHERE user_id = ? AND id NOT IN (
                    SELECT id FROM notifications
                    WHERE user_id = ?
                    ORDER BY created_at DESC
                    LIMIT ?
                )
                """, arguments: [userId, userId, keepCount])
        }
    }
}
